transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java

```diff
@@ -161,18 +162,36 @@ private void notifyHealthCheck(Future<Boolean> future, Channel ch, Promise<Chann
                     handler.channelAcquired(ch);
                     promise.setSuccess(ch);
                 } catch (Throwable cause) {
-                    closeAndFail(ch, cause, promise);
+                    Promise<Void> channelClosePromise = ch.eventLoop().newPromise();
+                    onUnhealthyChannelClosed(channelClosePromise);
+                    closeAndFail(ch, cause, promise, channelClosePromise);
                 }
             } else {
-                closeChannel(ch);
+                Promise<Void> channelClosePromise = ch.eventLoop().newPromise();
+                onUnhealthyChannelClosed(channelClosePromise);
+                closeChannel(ch, channelClosePromise);
                 acquire(promise);
             }
         } else {
-            closeChannel(ch);
+            Promise<Void> channelClosePromise = ch.eventLoop().newPromise();
+            onUnhealthyChannelClosed(channelClosePromise);
+            closeChannel(ch, channelClosePromise);
             acquire(promise);
         }
     }
 
+    private void onUnhealthyChannelClosed(Promise<Void> channelClosePromise) {
+        channelClosePromise.addListener(new GenericFutureListener<Future<? super Void>>() {
```

Comment:
the GenericFutureListener can be stored as instance in the class itself and re-used. Not need to create a new one all the time.